### PR TITLE
Add DeleteFunc & DeleteAll 

### PR DESCRIPTION
**Pull request by paddlesteamer**
I added my version of `Flush` method and a new method called `FlushWithFilter`. This is not a serious PR, I just thought you may be interested in the differences in my version.

The differences between our `Flush` methods are:
**1.** in my version, `Flush` returns flushed `map[string]Item`
**2.** and calls `onEvicted` on flushed items. 

`FlushWithFilter` flushes filtered items and returns them. It may not be fast in caches with big size since it iterates over every item in the cache but it's up to the user to use it.

### CODE REVIEW

**Comment by arp242** [comment]
Thanks for your PR :-)

One of the goals of the v1 release is to retain 100% compatibility with go-cache, so changing the behaviour of an existing function like `Flush()` is kind of a non-starter I'm afraid; it would break people's existing code.

More importantly, I mostly consider `Flush()` to mean "reset this database"; `Reset()` would probably have been a better name. I'm not sure if this is a good change on that account. What if I want to reset the database *without* calling OnEvicted? That would actually be quite hard now I think?

Right now you can already get this functionality with:

	items := c.Keys()  // Or use `Items()` if you want all the items. 
	for _, k := range items {
		c.Delete(k)
	}

We could *maybe* consider adding a `DeleteAll()` for this, but it would be useful to understand the exact use case for this? I'd prefer to avoid adding too much problem-specific API and favour "building blocks" API that people can use to do all sorts of things I never thought of, but if it solves a specific problem that would be hard to solve otherwise, yeah, we should do something about that one way or the other :-)

---

I think `FlushWithFilter()` is at the very least misnamed, as Flush means "reset" and not "delete" as mentioned above.

I included a [`DeleteFunc()`](https://github.com/zgoat/zcache/commit/e6b6776a44882be33d82f1f225cd4bac49314a8c) before, which is more or less the same as `FlushWithFilter()`, and removed it again the next day [when I added `Keys()`](https://github.com/zgoat/zcache/commit/b15ef52b2de4316e12bba5b57d1bf1beee2873bd). Like in the above example, it seems to me that you can get this already by writing a `for` loop, but perhaps there is a particular use case I didn't think of?

**Comment by paddlesteamer** [comment]
> Right now you can already get this functionality with:
>```
>items := c.Keys()  // Or use `Items()` if you want all the items. 
>for _, k := range items {
>	c.Delete(k)
>}
>```

This seems like it provides the same functionality but actually this may cause bugs. For example, let's say we have a cache contains key-value pairs `foo => 1` and `bar => 2` and let's consider this scenario:
- Call `c.Keys()` and get an array of `foo` and `bar`
- In the meantime, from another goroutine 'foo' is set to `3`
- We iterate over key array and delete this freshly set `foo`

But with my implementation, all items are returned and the cache is flushed. So the new `foo` key-value pair is untouched and we can do whatever we want with the returned items which include the old `foo` safely.

The same thing is also true for `FlushWithFilter/DeleteFunc`. It is very dangerous to call `Keys()` and then iterate over it without a mutex and it's not convenient for a user to do this with an explicit mutex.

But of course, it's your call to include this or not. Maybe keeping the library simple is better, I don't know. But if you decide to use these functions, I think your suggestion of namings is better and I can change the names accordingly. I'm always bad at finding names to functions and variables :(

And `onEvicted` was the reason I chose to use `go-cache` in the first place. In one of my projects, I keep track of some files with some metadata in the cache, and there comes a time it is necessary to reset the cache. But I need `onEvicted` to be called to remove those files from the filesystem. So I would prefer a `DeleteAll` method to delete all items with an`onEvicted` call afterward and a `DeleteFunc` method to choose what I want to delete and again with an `onEvicted` call afterward. 

Even if my case may be specific and rare, I think a lot of people would need `DeleteAll` with `onEvicted` because they would want a clean exit on their codes. Putting `defer c.DeleteAll()` after creating a new cache would help a lot of people.

To sum up, my suggestion is to rename `Flush` and `FlushWithFilter` to `DeleteAll` and to `DeleteFunc` in my commit and keep the good ol' `Flush` as it is.

**Comment by arp242** [comment]
Thanks for the details.

I'm not opposed to adding this, it's just that generally speaking small APIs are better IMHO, so wanted to ensure it actually solves a problem that's tricky to solve otherwise, rather than just "it's convenient for me in my particular use case" :-)

I'm fine with adding this, just change the names as mentioned before; the `*Func()` suffix is already used in the standard library (`strings.IndexFunc()`, `regex.ReplaceAllFunc()`, etc.) so it's probably best to stick to that instead of `WithFilter()`.

One thing my DeleteFunc() has is a second return value to tell it to stop looping, and perhaps it's also better to pass `Item` instead of `Item.Object` so you can check `Expiration` as well? Not sure about that last one, but I think it might make it a bit more flexible 🤔

> And onEvicted was the reason I chose to use go-cache in the first place.

Yeah, it's just that Flush() is the one exception where it shouldn't be called.

**Comment by paddlesteamer** [comment]
I agree with all of the things you said and I made changes accordingly. I'd like to know what you think. 

> One thing my DeleteFunc() has is a second return value to tell it to stop looping

I really liked this idea, by the way, it's a huge performance improvement. I wish native golang `*Func` methods would also have this.

> perhaps it's also better to pass `Item` instead of `Item.Object` so you can check `Expiration` as well?

This would cover the use cases that we can't think of and I think `onEvicted` should be the same too. I know you want `zcache` to be compatible with `go-cache` but maybe it is something to change in v2?

**Comment by arp242** [comment]
Sorry it took so long to get back to you; I was pretty sick for a few weeks and then had some things to catch up on 😅

> > perhaps it's also better to pass Item instead of Item.Object so you can check Expiration as well?

> This would cover the use cases that we can't think of and I think onEvicted should be the same too. I know you want zcache to be compatible with go-cache but maybe it is something to change in v2?

Sure, I've added it to the list of items here: https://github.com/zgoat/zcache/issues/1
